### ISSUE TITLE
Remote data packages with resources relative to their URL should be safe

### ISSUE THREAD

**Issue by vitorbaptista** [frictionlessdata/datapackage-py#63]
For example, the datapackage.json at https://raw.githubusercontent.com/datasets/gdp/master/datapackage.json defines a resource with a relative path (`data/gdp.csv`), which should be downloaded at https://raw.githubusercontent.com/datasets/gdp/master/data/gdp.csv. This should be considered safe, as their "base path" is the same (`https://raw.githubusercontent.com/datasets/gdp/master`).

This currently isn't the case. This issue was reported by @ndkv's on https://github.com/ckan/ckanext-datapackager/issues/38#issuecomment-199694709

> I ran into a similar safe/unsafe issue: uploading a remote `datapackage.json` (i.e. unzipped) that uses relative `path`s for its resources [1] gives an `Unsafe ...` error while uploading the same package with "absolute" `url`s [2] works fine. 
> 
> This seems to be the intended behaviour
> 
> > To avoid a resource with a path like `/etc/shadow`, this extension only accepts paths relative to the datapackage's ZIP file. As you haven't uploaded a ZIP, but the actual datapackage.json, that doesn't exist, so no local path is valid.
> 
> While I see the need to prevent `/etc/shadow`-like paths, I'm not sure disallowing relative resource `path`s for remote packages (i.e. unzipped) is intended as well. Also, the spec allows relative `path`s for remote pacakges
> 
> > NOTE: the use of a `url` allows a data package to reference data not contained locally in the Data Package. _The `path` attribute may also be used for Data Packages located online – in this case it determines the resource file’s URL relative to the `datapackage.json`’s URL._
> 
> I'm not familiar (at all) with inner working of CKAN/extension code but it seems that the culprit is line 151 in the `safe` function linked above
> 
> ``` python
> if not resource.local_data_path.startswith(self.base_path):
> ```
> 
> I guess remote data packages have their URL as `base_path`. Relative `path`s are rejected since they don't start with the URL. `url`s on the other hand pass this test just fine. 
> 
> So, to allow relative paths on remote packages we'd need to treat `path` differently, right? 
> 
> [1] https://github.com/ndkv/datasets/blob/master/gemeenschappelijke_regelingen/partial/datapackage.json
> [2] https://github.com/ndkv/datasets/blob/gh-pages/gemeenschappelijke_regelingen/partial/datapackage.json
> [3] http://dataprotocols.org/data-packages/#required-fields-1
> 
> BTW The "unsafe" message initially tricked me into thinking that the package's `description` attribute was the culprit as it contains links (it isn't and the links are stripped). It might be better, if possible, to mention something about the difference between `url` and `path`.

**Comment by vitorbaptista** [comment]
We currently consider any remote resource as safe, but what about a remote datapackage.json like:

```
{
    "name": "unsafe",
    "resources": [
        { "path": "../../foo.csv" }
    ]
}
```

Say that the datapackage.json lives at `http://www.foo.com/bar/baz/datapackage.json`, then the resource would be at `http://www.foo.com/foo.csv`. Is it unsafe or not?

Considering that this datapackage is equivalent to:
```
{
    "name": "unsafe",
    "resources": [
        { "path": "http://www.foo.com/foo.csv" }
    ]
}
```

and that is safe, I say we consider the first one safe as well. All in all, we consider every remote resource to be safe.

Thoughts?

**Comment by ndkv** [comment]
While researching this issue I stumbled on #97. 

Once that is fixed `safe()` should be updated to check `remote_data_path`s as well. [Line 151] (https://github.com/frictionlessdata/datapackage-py/blob/master/datapackage/datapackage.py#L151) should be replaced by something like this.

```python
if resource.local_data_path:
    if not resource.local_data_path.startswith(self.base_path):
        return False

if resource.remote_data_path:
    if not resource.remote_data_path.startswith(self.base_path):
        return False
```

Thoughts? I'll send pull requests of you think this is the right approach.

**Comment by vitorbaptista** [comment]
I think this needs more discussion.

We currently consider all remote resources to be safe. The issue with this is when uploading a datapackage to, for example, a CKAN instance using `ckanext-datapackager`. The CKAN instance will load the datapackage and its resources if it's safe. We only consider local resources to be safe it they're relative to the `datapackage.json` path, so a user couldn't send a datapackage with a resource as `/etc/shadow`, for example. However, as all remote resources are considered safe, they could still send a datapackage pointing to `http://10.0.0.1/secret-file.txt` in the server's local network and have its contents published on CKAN.

This is kind of difficult to exploit, as it requires knowledge of the server's intranet and the ability to upload a datapackage to CKAN. This is only considering CKAN, though. There might be uses of `datapackage-py` that makes this hole easier to exploit.

We could limit remote resources to the same base path of the `datapackage.json`, the same way we do with local resources, but that means that it won't be possible for a datapackage to point to resources outside of it domain (e.g. the World Bank data portal).

I think it's OK to consider all remote resources as safe, as we do now, maybe just documenting this possible attack vector so clients can handle it. What do you think, @pwalsh @roll @rgrp?

**Comment by ndkv** [comment]
Another option is to detect local network IPs with e.g. the [ipaddress](https://pypi.python.org/pypi/ipaddress) module. We can then either flag them as unsafe or issue a warning in case someone is willingly/knowingly working with private IPs.

**Comment by roll** [comment]
`ipaddress` looks like a real complication.. 

Sometimes such kind of problems could be fixed stepping a few steps back and re-checking everything starting from `is this a properly defined problem we try to solve` or may be we could solve other problem to achieve the real goal? 

I'm not fully into this situation so a a fresh look - why e.g. just to do not forbid dots in paths?

**Comment by vitorbaptista** [comment]
@roll The local paths are fine: if it's inside the `datapackage.json`'s folder, it's safe; otherwise it's unsafe. For remote paths is a bit more tricky: there can be URLs only accessible from the server.

The attack vector I'm thinking is a user with permission to upload to a CKAN instance using `ckanext-datapackager` (e.g. http://datahub.io), and a server with access to a webserver that the user can't access directly (e.g. in a Intranet, but can be outside). The user could upload a `datapackage.json` with a remote resource that points to this URL. CKAN would then hit that URL, download whatever is in it and upload as a CKAN resource, which can be read by the attacker.

This is a very narrow attack vector without a simple solution. In this case, I'd suggest we should just log it in a comment (e.g. the `safe()` docstring) letting the user know, so she can evaluate and decide what to do about it (if anything at all). At most, we could add a `safe(no_remote_resources=False)` or similar to allow the user to decide if remote URLs are OK or not, but I wouldn't implement that before we have an actual user needing it (probably the datahub itself).

**Comment by roll** [comment]
FIXED by #139 and #140
